metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/DirectSqlTable.java

```diff
@@ -0,0 +1,217 @@
+        }
+        ResultSetExtractor<Map<String, List<QualifiedName>>> handler = rs -> {
+            final Map<String, List<QualifiedName>> result = Maps.newHashMap();
+            while (rs.next()) {
+                final String schemaName = rs.getString("schema_name");
+                final String tableName = rs.getString("table_name");
+                final String uri = rs.getString("location");
+                final List<QualifiedName> names = result.computeIfAbsent(uri, k -> Lists.newArrayList());
+                names.add(QualifiedName.ofTable(catalogName, schemaName, tableName));
+            }
+            return result;
+        };
+        try {
+            return jdbcTemplate.query(queryBuilder.toString(), params.toArray(), handler);
+        } finally {
+            this.fastServiceMetric.recordTimer(
+                HiveMetrics.TagGetTableNames.getMetricName(), registry.clock().wallTime() - start);
+        }
+    }
+
+    /**
+     * Locks the iceberg table for update so that no other request can modify the table.
+     * @param tableId table internal id
+     */
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void lockIcebergTable(final Long tableId) {
+        Boolean isLocked = null;
+        try {
+            isLocked = jdbcTemplate.queryForObject(SQL.ICEBERG_TABLE_LOCK,
+                new SqlParameterValue[] {new SqlParameterValue(Types.BIGINT, tableId),
+                    new SqlParameterValue(Types.VARCHAR, PARAM_METADATA_LOCK), }, Boolean.class);
+        } catch (EmptyResultDataAccessException ignored) { }
+        if (isLocked == null) {
+            insertTableParam(tableId, PARAM_METADATA_LOCK, 1);
+        } else if (isLocked) {
+            throw new IllegalStateException("Iceberg table is locked.");
+        } else {
+            updateTableParam(tableId, PARAM_METADATA_LOCK, 1);
+        }
+    }
+
+    /**
+     * Unlocks the iceberg table after update so that another request can modify the table.
+     * @param tableId table internal id
+     */
+    public void unlockIcebergTable(final Long tableId) {
+        updateTableParam(tableId, PARAM_METADATA_LOCK, 0);
+    }
+
+    private void insertTableParam(final Long tableId, final String parameter, final int value) {
+        jdbcTemplate.update(SQL.INSERT_TABLE_PARAMS, new SqlParameterValue(Types.BIGINT, tableId),
+            new SqlParameterValue(Types.VARCHAR, parameter), new SqlParameterValue(Types.VARCHAR, value));
+    }
+
+    private void updateTableParam(final Long tableId, final String parameter, final int value) {
+        jdbcTemplate.update(SQL.UPDATE_TABLE_PARAMS, new SqlParameterValue(Types.VARCHAR, value),
+            new SqlParameterValue(Types.BIGINT, tableId), new SqlParameterValue(Types.VARCHAR, parameter));
+    }
+
+    /**
+     * Returns the table internal id.
+     * @param tableName table name
+     * @return table id
+     */
```

Comment:
Should this include @Transactional(readOnly = true)?

Reply:
yes.